x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/scalar/datetime/CurrentDateTime.java

```diff
@@ -7,63 +7,38 @@
 package org.elasticsearch.xpack.sql.expression.function.scalar.datetime;
 
 import org.elasticsearch.xpack.sql.expression.Expression;
-import org.elasticsearch.xpack.sql.expression.function.scalar.ConfigurationFunction;
 import org.elasticsearch.xpack.sql.session.Configuration;
-import org.elasticsearch.xpack.sql.tree.Source;
 import org.elasticsearch.xpack.sql.tree.NodeInfo;
+import org.elasticsearch.xpack.sql.tree.Source;
 import org.elasticsearch.xpack.sql.type.DataType;
 
 import java.time.ZonedDateTime;
-import java.util.Objects;
 
-public class CurrentDateTime extends ConfigurationFunction {
+public class CurrentDateTime extends CurrentFunction {
+
     private final Expression precision;
-    private final ZonedDateTime dateTime;
 
     public CurrentDateTime(Source source, Expression precision, Configuration configuration) {
-        super(source, configuration, DataType.DATETIME);
+        super(source, configuration, nanoPrecision(configuration.now(), precision), DataType.DATETIME);
         this.precision = precision;
-        int p = precision != null ? ((Number) precision.fold()).intValue() : 0;
-        this.dateTime = nanoPrecision(configuration().now(), p);
     }
 
-    @Override
-    public Object fold() {
-        return dateTime;
+    Expression precision() {
+        return precision;
     }
 
     @Override
     protected NodeInfo<CurrentDateTime> info() {
         return NodeInfo.create(this, CurrentDateTime::new, precision, configuration());
     }
 
-    @Override
-    public int hashCode() {
-        return Objects.hash(dateTime);
-    }
-
-    @Override
-    public boolean equals(Object obj) {
-        if (this == obj) {
-            return true;
-        }
-
-        if (obj == null || getClass() != obj.getClass()) {
-            return false;
-        }
-
-        CurrentDateTime other = (CurrentDateTime) obj;
-        return Objects.equals(dateTime, other.dateTime);
-    }
-
-    static ZonedDateTime nanoPrecision(ZonedDateTime zdt, int precision) {
-        if (zdt != null) {
-            int nano = zdt.getNano();
-            if (precision >= 0 && precision < 10) {
-                // remove the remainder
-                nano = nano - nano % (int) Math.pow(10, (9 - precision));
-                return zdt.withNano(nano);
-            }
+    static ZonedDateTime nanoPrecision(ZonedDateTime zdt, Expression precisionExpression) {
+        int precision = precisionExpression != null ? ((Number) precisionExpression.fold()).intValue() : 0;
```

Comment:
Use `Foldables.intValueOf` instead as it triggers a full conversion and also checks the foldability of the expression.